### ISSUE TITLE
`Mail.new STDIN.read` raises `invalid byte sequence in US-ASCII`

### ISSUE THREAD

**Issue by leafac** [mikel/mail#1348]
The stack trace:

```
invalid byte sequence in US-ASCII
mail-2.7.1/lib/mail/message.rb:2032:in `match'
mail-2.7.1/lib/mail/message.rb:2032:in `match'
mail-2.7.1/lib/mail/message.rb:2032:in `set_envelope_header'
mail-2.7.1/lib/mail/message.rb:2132:in `init_with_string'
mail-2.7.1/lib/mail/message.rb:137:in `initialize'
mail-2.7.1/lib/mail/mail.rb:51:in `new'
mail-2.7.1/lib/mail/mail.rb:51:in `new'
```

The locale on the machine:

```console
$ locale
LANG=C.UTF-8
LANGUAGE=
LC_CTYPE="C.UTF-8"
LC_NUMERIC="C.UTF-8"
LC_TIME="C.UTF-8"
LC_COLLATE="C.UTF-8"
LC_MONETARY="C.UTF-8"
LC_MESSAGES="C.UTF-8"
LC_PAPER="C.UTF-8"
LC_NAME="C.UTF-8"
LC_ADDRESS="C.UTF-8"
LC_TELEPHONE="C.UTF-8"
LC_MEASUREMENT="C.UTF-8"
LC_IDENTIFICATION="C.UTF-8"
LC_ALL=
```

The Exim configuration which pipes the email to a Ruby script: https://github.com/leafac/www.kill-the-newsletter.com/blob/master/exim.conf

The Ruby script itself: https://github.com/leafac/www.kill-the-newsletter.com/blob/master/mail_handler.rb

I can provide an example of an email that causes the problem via a private channel.

I read a bunch about this online and I still can’t figure it out.

Should I change the locale to `en_US.utf8` by changing `/etc/environment`?

```
LANG=en_US.utf8
LC_ALL=en_US.UTF-8
```

Should I start my Ruby script with something like the following?

```ruby
Encoding.default_external = Encoding::UTF_8
Encoding.default_internal = Encoding::UTF_8
```

Should I invoke Ruby with the `--encoding` argument? I tried `--encoding UTF-8:UTF-8` but it didn’t work.

Should I do something else entirely?

Thank you for this gem and for your help.

**Comment by jeremy** [comment]
Classic problem with Ruby picking up US-ASCII from your environment. You can set LANG/LC_ALL globally, pass it per script invocation in your Exim config, or set `Encoding.default_external = …` in your script itself. Nice setup; good luck!

**Comment by leafac** [comment]
### First Attempt

Edit the `mail_handler.rb` to begin with

```ruby
Encoding.default_external = Encoding::UTF_8
```

or

```ruby
Encoding.default_external = Encoding::UTF_8
Encoding.default_internal = Encoding::UTF_8
```

In both cases the problem persists (with UTF-8, this time): `invalid byte sequence in UTF-8`.

I expected that, because I think that using the `--encoding` command-line flag when invoking Ruby, which had I tried before, amounts to the same as adding the lines above to the beginning of the script.

### Second Attempt

Undid the **First Attempt**.

Change the locale in the operating system with

```console
# update-locale LANG=en_US.UTF-8
```

and rebooted such that now

```console
# locale
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
```

Surprisingly, the message is still `invalid byte sequence in US-ASCII`, as opposed to `UTF-8`.

### Third Attempt

Set `LC_ALL` as well such that

```console
# cat /etc/default/locale
LANG=en_US.UTF-8
LC_ALL=en_US.UTF-8
```

and rebooted such that now

```console
# locale
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=en_US.UTF-8
```

Same: `invalid byte sequence in US-ASCII`.

### Fourth Attempt

With the locale as in the **Third Attempt**, recompiled Ruby with `ruby-install`.

No dice: `invalid byte sequence in US-ASCII`.

### Conclusion

Please help 🙏

**Comment by leafac** [comment]
[Here’s](https://github.com/leafac/www.kill-the-newsletter.com/commit/f4996a5080948074110436b9fde987bac2b41a1c) what’s doing it for me for now:

1. I configured Exim to pass the `LANG` environment variable.†
2. I `String#scrub` the data as it comes from `STDIN`.

I believe I’m losing characters that are encoded in anything that isn’t UTF-8. But it seems infrequent enough that my users probably won’t care too much. Also, I’m not sure if there’s anything I can do about it.

† I ended up using `LANG=C.UTF-8`, as opposed to the `LANG=en_US.UTF-8` that I used in the comments above. I was under the wrong impression that `LANG=en_US.UTF-8 ⇒ Encoding.default_external = Encoding::UTF_8` and `LANG=C.UTF-8 ⇒ Encoding.default_external != Encoding::UTF_8`. So I was fighting the `LANG=C.UTF-8`. But I was wrong: `LANG=C.UTF-8 ⇒ Encoding.default_external == Encoding::UTF_8`. The problem seems to be that Exim was stripping the `LANG=C.UTF-8` from the environment. In other Ruby processes, for example, the Sinatra server, the encoding was `Encoding.default_external = Encoding::UTF_8` just fine.
